agent/consul/state/catalog.go

```diff
@@ -4628,6 +4690,67 @@ func (s *Store) combinedServiceNodesTxn(tx ReadTxn, ws memdb.WatchSet, names []s
 	return maxIdx, resp, nil
 }
 
+func upstreamsFromGWTxn(tx ReadTxn, service structs.ServiceName) ([]structs.ServiceName, error) {
+	val, err := tx.First(tableConfigEntries, indexID, configentry.KindName{Kind: structs.BoundAPIGateway, Name: service.Name})
+	if err != nil {
+		return nil, err
+	}
+
+	if gw, ok := val.(*structs.BoundAPIGatewayConfigEntry); ok {
+		serviceIDs := gw.ListRelatedServices()
+		names := make([]structs.ServiceName, 0, len(serviceIDs))
+		for _, id := range serviceIDs {
+			names = append(names, structs.NewServiceName(id.ID, &id.EnterpriseMeta))
+		}
+		return names, nil
+	}
+
+	return nil, errors.New("not an APIGateway")
+}
+
+// downstreamsForServiceTxn will find all downstream services that could route traffic to the input service.
+// There are two factors at play. Upstreams defined in a proxy registration, and the discovery chain for those upstreams.
+func (s *Store) downstreamGatewaysForServiceTxn(tx ReadTxn,  service structs.ServiceName) (uint64, []structs.ServiceName, error) {
+	iter, err := tx.Get(tableConfigEntries, indexLink, service.ToServiceID())
+	if err != nil {
+		return 0, nil, err
+	}
+
+	var (
+		idx  uint64
+		resp []structs.ServiceName
+		seen = make(map[structs.ServiceName]struct{})
+	)
+	for raw := iter.Next(); raw != nil; raw = iter.Next() {
+		entry, ok := raw.(*structs.BoundAPIGatewayConfigEntry)
+		if !ok {
+			continue
+		}
+
+		if entry.ModifyIndex > idx {
+			idx = entry.ModifyIndex
+		}
+
+		gwServiceName := structs.NewServiceName(entry.Name, &entry.EnterpriseMeta)
+		if _, ok := seen[gwServiceName]; ok {
+			continue
+		}
+
+		// This is a hack because for some reason memdb isn't dropping the bound api gateway from the link index after
+		// removing a service from the Services field
```

Comment:
This may be worth running by some of the other core folks. I can take a look to see if I can figure out why the index isn't getting rebuilt, but this does seems problematic if it's the case since it could (potentially) result in an underlying memory leak where we're holding stale index data even when resources are updated/don't exist anymore.

Reply:
ran it by them, issue was a missing `DeepCopy` in the controller code which we patched with https://github.com/hashicorp/consul/pull/19657/commits/f613a351c434e6cd0b60d363b4e3fe31424ec636 commit